src/hotspot/share/gc/g1/g1ConcurrentRefineThread.hpp

```diff
@@ -84,4 +84,30 @@ class G1ConcurrentRefineThread: public ConcurrentGCThread {
   double vtime_accum() { return _vtime_accum; }
 };
 
+// Singleton special refinement thread, registered with the dirty card queue.
+// This thread supports notification of increases to the number of cards in
+// the dirty card queue, which may trigger activation of this thread when it
+// is not already running.
+class G1PrimaryConcurrentRefineThread final : public G1ConcurrentRefineThread {
+  Semaphore _notifier;
+  DEFINE_PAD_MINUS_SIZE(0, DEFAULT_CACHE_LINE_SIZE, 0);
+  volatile size_t _threshold;
```

Comment:
Might make sense to add some additional comment for the threshold, similar to the description you have in the PR summary: 
```
The primary thread uses an atomic activation threshold that also serves as "is
running" state. This activation threshold is used by the write barrier
support to determine whether the thread's semaphore needs to be signaled.
```

Reply:
I meant to do that, and then forgot.  Thanks for point it out.  I'll add it before pushing.